src/conn.rs

```diff
@@ -451,9 +458,56 @@ impl Queryable for Store {
 }
 
 impl Syncable for Store {
+    // TODO refactor this out of there, somehow!
     fn sync(&mut self, server_uri: &String, user_uuid: &String) -> Result<()> {
         let uuid = Uuid::parse_str(&user_uuid)?;
-        Ok(Syncer::flow(&mut self.sqlite, server_uri, &uuid)?)
+        let mut tx_list = None;
+        {
+            let mut db_tx = self.sqlite.transaction()?;
+            if let Some(list) = Syncer::flow(&mut db_tx, server_uri, &uuid)? {
+                tx_list = Some(list);
+            }
+
+            // TODO this should be done _after_ any parts below are transacted!!
+            // Commits any changes Syncer made (tu mappings during an upload!)
+            db_tx.commit()?;
+        }
+        let mut last_tx_entid = None;
+        let mut last_tx_uuid = None;
+        // Giant hack: we're expecting 'parts' to not actually contain metadata
+        // about any transactions - just their datoms! we generate metadata ourselves
+        // while transacting, thus loosing when transactions were made originally /o\
+        match tx_list {
+            Some(list) => {
+                for tx in list {
+                    let in_progress = self.begin_transaction()?;
```

Comment:
I think you should lift this out of the `for` loop — you can transact multiple times against an `in_progress`, and they'll all share a single SQLite transaction.